src/bento/components/bento-autocomplete/1.0/component.tsx

```diff
@@ -0,0 +1,555 @@
+        if (slot) {
+          possibleElements = querySelectorAllInSlot(slot, inputSelector);
+        }
+      }
+      if (possibleElements.length !== 1) {
+        onError(`should contain exactly one <input> or <textarea> descendant.`);
+        return;
+      }
+      return possibleElements[0];
+    },
+    [onError]
+  );
+
+  const setupInputElement = useCallback(
+    (element: HTMLElement) => {
+      const inputElement = getSingleInputOrTextarea(element);
+      if (inputElement) {
+        if (inputElement.hasAttribute('type')) {
+          const inputType = inputElement.getAttribute('type');
+          if (inputType !== 'text' && inputType !== 'search') {
+            onError(
+              `requires the "type" attribute to be "text" or "search" if present on <input>.`
+            );
+          }
+        }
+        inputElement.classList.add(classes.input);
+        inputElement.setAttribute('dir', 'auto');
+        inputElement.setAttribute('aria-autocomplete', 'both');
+        inputElement.setAttribute('aria-controls', containerId.current!);
+        inputElement.setAttribute('aria-haspopup', 'listbox');
+        inputElement.setAttribute('aria-owns', containerId.current!);
+        inputElement.setAttribute('aria-expanded', 'false');
+        if (inputElement.tagName === 'INPUT') {
+          inputElement.setAttribute('role', 'combobox');
+          inputElement.setAttribute('aria-multiline', 'false');
+        } else {
+          inputElement.setAttribute('role', 'textbox');
+        }
+
+        inputRef.current = inputElement as HTMLInputElement;
+      }
+    },
+    [getSingleInputOrTextarea, onError, classes.input, containerId]
+  );
+
+  const validateProps = useCallback(() => {
+    if (!isValidFilterType(filter)) {
+      onError(`Unexpected filter: ${filter}.`);
+    }
+    if (!inline && suggestFirst && filter !== 'prefix') {
+      onError(`"suggest-first" expected "filter" type "prefix".`);
+      setShouldSuggestFirst(false);
+    }
+  }, [filter, onError, inline, suggestFirst]);
+
+  const showAutocompleteResults = useMemo(() => {
+    if (!areResultsDisplayed || data?.length === 0) {
+      return false;
+    }
+    return substring.length >= minChars;
+  }, [data, substring, minChars, areResultsDisplayed]);
+
+  const truncateToMaxItems = useCallback(
+    (data: Item[] = []) => {
+      if (maxItems && maxItems < data.length) {
+        return data.slice(0, maxItems);
+      }
+      return data;
+    },
+    [maxItems]
+  );
+
+  const filteredData = useMemo(() => {
+    const normalizedValue = substring.toLocaleLowerCase();
+
+    const allFilteredData = data?.filter((item: Item) => {
+      if (typeof item === 'object') {
+        item = getValueForExpr(item, filterValue);
+      }
+      if (typeof item !== 'string') {
+        onError(`data property "${filterValue}" must map to string type.`);
+        return;
+      }
+      item = item.toLocaleLowerCase();
+
+      switch (filter) {
```

Comment:
If I was a react/preact user of this library I think I'd like to give this component a filter function and it just invokes that function. Which would allow me (as the consumer) to customize how the list is filtered.

I realize for the bento web-component and amp...that probably isn't as possible, or at least not as easy. So I guess we should just leave this as is.